### PR TITLE
Wrote my take-aways from debugging a memory leak

### PR DESCRIPTION
**Pull request by kraenhansen**
## What, How & Why?

I hope this may help anyone trying to debug memory leaks in our repository.
It's a very simplified and minimalistic guide.

I've tagged a lot of you for review, feel free to skip this though 🙂 
Read if this might be interesting for you or have prior experience with debugging memory leaks.

### CODE REVIEW

File: contrib/guide-memory-leaks.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+You're debugging an issue where using Realm JS takes up lot of memory, perhaps it grows linearly as the program runs and perhaps the process exits with out of memory exceptions.
+
+Generally speaking memory leaks will fall into one of two categories:
+
+- JavaScript objects referencing each other in a way that prevents garbage collection.
+- Everything else.
+
+## Leaking JavaScript objects
+
+There's a couple of reasons this guide won't dwelve too much on this type of leak:
```
Comment (fronck) [inline contrib/guide-memory-leaks.md:12]: [spelling]  *"dwell on" or "delve into"

File: contrib/guide-memory-leaks.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+- JavaScript is garbage collected, which limits the risk of the most obvious mistakes.
+- JavaScript is a minority language in our codebase.
+- This is not where we've had leaks historically.
+
+### Running a Node.js script (attaching with Chrome dev-tools)
+
+Since both Node.js and Chrome are using the V8 JavaScript engine, we can use the Chromium dev-tools to inspect the process on any of our supported development platforms.
+
+Run the process with the `--inspect` (or `--inspect-brk` to pause the process until the debugger attach) runtime parameter:
```
Comment (fronck) [inline contrib/guide-memory-leaks.md:21]:
```suggestion
Run the process with the `--inspect` (or `--inspect-brk` to pause the process until the debugger attaches) runtime parameter:
```

File: contrib/guide-memory-leaks.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+```
+node --inspect index.js
+```
+
+Navigate a Chromium browser to and click the "inspect" link next to the target representing the Node.js process
```
Comment (fronck) [inline contrib/guide-memory-leaks.md:27]: Navitage a Chromium browser to where?

File: contrib/guide-memory-leaks.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+```
+chrome://inspect
+```
+
+Besides access to the console, ability to set breakpoints and execute code step-by-step, this allows you to profile memory allocation by taking and comparing snapshots of the JavaScript heap.
+
+## Everything else ...
+
+Majority of our code is C++ code wrapping the Realm Object Store to expose object and property access. This is probably the part of our repository most ripe for memory leaks.
```
Comment (fronck) [inline contrib/guide-memory-leaks.md:37]:
```suggestion
The majority of our code is C++ code wrapping the Realm Object Store to expose object and property access. This is probably the part of our repository most ripe for memory leaks.
```

File: contrib/guide-memory-leaks.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+### Running a Node.js script or React Native App (attaching with Instruments for macOS)
+
+Instruments can either attach to a running process (on the host or a simulator) or you can "Choose Target ..." to have Instruments start and run the app (making it possible to attach from the beginning).
+
+An advantage is that it provides a great GUI. Disadvantage is that it cannot attach to all processes (ex Android process running in a simulator) and that it's only available on macOS.
```
Comment (fronck) [inline contrib/guide-memory-leaks.md:43]:
```suggestion
An advantage is that it provides a great GUI. Disadvantage is that it cannot attach to all processes (e.g., Android process running in a simulator) and that it's only available on macOS.
```

File: contrib/guide-memory-leaks.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+### Running a Node.js script (taking heap snapshots with Google Performance Tools)
+
+[Google Performance Tools / gperftools](https://github.com/gperftools/gperftools) is a library that either gets statically linked to a program or dynamically loaded at runtime.
+
+> gperftools is a collection of a high-performance multi-threaded malloc() implementation, plus some pretty nifty performance analysis tools.
+
+gperftools can be installed using brew.
```
Comment (fronck) [inline contrib/guide-memory-leaks.md:51]:
```suggestion
On MacOS, gperftools can be installed using brew.
```

File: contrib/guide-memory-leaks.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+pprof --svg --base=heapprof.0001.heap --add_lib=node_modules/realm/compiled/napi-v4_darwin_x64/realm.node $(which node) heapprof.0002.heap > perf.svg
+```
+
+Note: You might need to install the Graphviz tool as well (running `brew install gv` on macOS).
+
+The SVG can be inspected to get an idea of the flow of allocations and hopefully locate the source of any memory leaks.
+
+### Running a Node.js script using Valgrind
+
+Some might say that Valgrind is not exactly the "new kid on the block" and as such has a UI and output which is harder to wrap your head around. Valgrind can be installed [via a brew tab](https://github.com/LouisBrunner/valgrind-macos) on macOS.
```
Comment (fronck) [inline contrib/guide-memory-leaks.md:77]:
```suggestion
Some might say that Valgrind is not exactly the "new kid on the block" and as such has a UI and output which is harder to wrap your head around. Valgrind can be installed [via a brew tap](https://github.com/LouisBrunner/valgrind-macos) on macOS. 
```

**Review by fronck — CHANGES_REQUESTED** [review]
Minor grammar and typo corrections.

Otherwise:  nice :)

**Review by cesarvr — APPROVED** [review]

**Review by kneth — APPROVED** [review]
Nice :-)

**Review by fronck — APPROVED** [review]
👍

**Review by steffenagger — APPROVED** [review]

**Pull request merged**